pkg/vm/vm_core.go

```diff
@@ -1,15 +1,70 @@
 package vm
 
-import "github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+import (
+	"errors"
+
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
 
 // VirtualMachine represents the Cairo VM.
 // Runs Cairo assembly and produces an execution trace.
 type VirtualMachine struct {
-	RunContext  RunContext
-	currentStep uint
-	Segments    memory.MemorySegmentManager
+	RunContext     RunContext
+	currentStep    uint
+	Segments       memory.MemorySegmentManager
+	Trace          []TraceEntry
+	RelocatedTrace []RelocatedTraceEntry
 }
 
 func NewVirtualMachine() *VirtualMachine {
-	return &VirtualMachine{Segments: *memory.NewMemorySegmentManager()}
+	segments := memory.NewMemorySegmentManager()
+	trace := make([]TraceEntry, 0)
+	relocatedTrace := make([]RelocatedTraceEntry, 0)
+	return &VirtualMachine{Segments: *segments, Trace: trace, RelocatedTrace: relocatedTrace}
+}
+
+func (v *VirtualMachine) RelocateTrace(relocationTable *[]uint) error {
+	if len(*relocationTable) < 2 {
+		return errors.New("No relocation found for execution segment")
+	}
+	segment1Base := (*relocationTable)[1]
+
+	for _, entry := range v.Trace {
+		v.RelocatedTrace = append(v.RelocatedTrace, RelocatedTraceEntry{
+			Pc: entry.Pc.RelocateAddress(relocationTable),
+			Ap: entry.Ap.RelocateAddress(relocationTable) + segment1Base,
+			Fp: entry.Fp.RelocateAddress(relocationTable) + segment1Base,
```

Comment:
Isn't this adding the segment1Base twice? (given that it is already added in `RelocateAddress`)

Reply:
I can't find where it's added, because `RelocateAddress` only returns `relocationTable[segmentIndex] + offset`.

Reply:
The `relocationTable` values have added the first address of the segment 0, which is `1`. In this case, I added to `Ap` and `Fp` the first address of the segment 1.

Reply:
`Ap` & `Fp` should always have segment index 1, I think the test values might be the problem here

Reply:
Remember when porting tests from cairo-vm, that segment_index is asumed to be 0 for pc and 1 for ap and fp

Reply:
Ok, so in this case if we take the trace entries from cairo-vm tests, we should add 1 to the ap and fp of the expected entries

Reply:
When creating the initial entry (aka the non relocated trace) you should convert each register to Relocatable doing this conversion:
cairo-vm (rust)
TraceEntry {
  pc: pc_offset
  ap: ap_offset
  fp: fp_offset
  }
  
  cairo-vm.go
  TraceEntry {
  pc: 0: pc_offset
  ap : 1: ap_offset
  fp: 1: fp_offset
  }
  Which is adding the correct segment index for each register, that is omitted in the cairo-vm trace for perfromance.
  Values in the RelocatedTrace should be left untouched